### PR TITLE
Fixed application generation (partially)

### PR DESCRIPTION
**Pull request by MathieuAA**
- Please make sure the below checklist is followed for Pull Requests.

- [ ] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
- [X] Tests are added where necessary
- [ ] Documentation is added/updated where necessary
- [X] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

File: generators/import-jdl/index.js [hunk 1]
```diff
@@ -100,50 +103,22 @@ module.exports = class extends BaseGenerator {
 
             parseJDL() {
                 this.log('The JDL is being parsed.');
+                const jdlImporter = new jhiCore.JDLImporter(this.jdlFiles, {
+                    databaseType: this.prodDatabaseType,
```
Comment (deepu105) [inline generators/import-jdl/index.js]: for app generation these wont be present is that fine?

File: generators/import-jdl/index.js [hunk 2]
```diff
@@ -154,93 +129,85 @@ module.exports = class extends BaseGenerator {
...
+                        generateApplicationFiles(
+                            this,
+                            this.importState.exportedApplications[0]
+                        );
+                    } catch (error) {
+                        this.error(`While generating applications from the parsed JDL\n${error}`);
+                    }
+                } else {
+                    // sub-folder generation, not yet handled
+                    this.applicationsLeftToGenerate = this.importState.exportedApplications
```
Comment (deepu105) [inline generators/import-jdl/index.js]: what is this usecase?

Comment (MathieuAA) [reply generators/import-jdl/index.js]: You could say it's temporary. Right now, the standard way of importing works. This is just to warn the user that generating entities/apps from one folder to another is not yet supported (I don't know if it's possible)

File: generators/import-jdl/index.js (hunk 1 above)
Comment (MathieuAA) [reply generators/import-jdl/index.js]: Why?

Comment (MathieuAA) [reply generators/import-jdl/index.js]: You mean the variables?

Comment (MathieuAA) [reply generators/import-jdl/index.js]: Oh I get it. It's perfectly fine. The importer will handle it just fine.

**Comment by MathieuAA** [comment]
My chief concern here is the composability of the import-jdl subgen with the app & entity subgens.
If you try to import one application & _n_ entities, it will fail because:
  - I think I don't call the app subgen the right way
  - The app & entity subgens are concurrent when composed... and we actually want the app subgen to be the one executed (launched, processed, finished) before the other

File: generators/import-jdl/index.js (hunk 2 above)
Comment (deepu105) [reply generators/import-jdl/index.js]: lets not support that at all, thats why I wanted to remove the path option in new JDL as well, as this could be easily abused and adds a lot of edge cases and possible bugs, lets keep it simple. Also it would be an issue in JHipster online if we support generating apps in other folders, our users are already familiar with the restrictions of yeoman and hence lets just stick to that

**Comment by deepu105** [comment]
@MathieuAA I might be able to help with composing, I have done my share of experiments with it for blueprints and stuff(and I added the original compose behaviours in the generators) so next week i'll spent some time to fix the blueprint issues, so i'll also try this out

File: generators/import-jdl/index.js (hunk 2 above)
Comment (MathieuAA) [reply generators/import-jdl/index.js]: Oh wait. I may have just understood the comment. I strongly disagree. Generating multiple apps is a must-have. Just so that we are on the same page here
  - there is no app path anymore
  - the chosen way to generate more than one app (instead of having a ton of JDL files) is to create folders so that any subfolder will be at `./<SUBFOLDER_i>` (the generator already handles not having paths as name)
  - a security would be not to generate anything if a baseName is an existing file (one can easily check this with the path module)

If we don't call the subgens in any of the subfolders, that's completely fine by me (that's why I wanted to inform the users of the remaining apps & entities to generate). But not supporting creating these subfolders with `.yo-rc.json` and `.jhipster` files in them would be a mistake in my opinion.

Comment (jdubois) [reply generators/import-jdl/index.js]: @MathieuAA I understand, but for JHipster Online we need to re-think the whole system, and that's just not possible now

Comment (deepu105) [reply generators/import-jdl/index.js]: ok, may be we were talking about different things here. Of course we have to support generating multiple apps form the same JDL, but then the only real use case for that is microservices and I dont see why we need to support that for a monolith, as IMO for monolith it would be a bad practice to have 2 apps on the same JDL.

Good to know that the path option is removed that means we have control of deciding where the path would be so I imagine the below use cases.

1. Entites in a JDL
  - same behavior as of today
2. single monolith app in a JDL with or without entities
  - `mkdir myApp && cd myApp`
  - Execute `jhipster import-jdl app.jh` and we create the below inside myApp folder
  - -- .yo-rc.json
    -- .jhipster
  - We call the app generator by specifying --with-entities flag
  - app is created in the `myApp` folder

3. Multiple microservice in a JDL with entities (say gateway, ms1, ms2)
  - `mkdir myMsApp && cd myMsApp`
  - Execute `jhipster import-jdl app.jh` and we create the below inside myApp folder
  - -- gateway
        -- .yo-rc.json
        -- .jhipster
    -- ms1
        -- .yo-rc.json
        -- .jhipster
    -- ms2
        -- .yo-rc.json
        -- .jhipster
  - We call the app generator by specifying --with-entities flag inside each sub folder, i.e `gateway, ms1 and ms2`
  - app is created in the sub folders accordingly

Now for case 3 the way yeoman works might be a challenge, but we have 2 options for that
1. Since we use our own CLI, we could cd into the folders we want and then execute the generators in the folders. This means import-jdl will not be a sub generator anymore and will be a CLI feature instead (to end user it doesnt matter as they will use it in the sameway, but a sideeffect would be that people wont be able to do `yo jhipster:import-jdl` anymore
2. Do some hack to cd into the folders before composing the app generators and cd out once done

Comment (MathieuAA) [reply generators/import-jdl/index.js]: The first option looks good, but it implies having to work a bit to achieve that.
Let's not bother ourselves about how we generate these multiple apps. Anyone can temporarily `cd` inside the folders and do a `yo jhipster`. I'd rather have the app composition working.
I can, in the meantime, issue a warning that a JDL file contains multiple apps (to inform it's a bad practise).

File: generators/import-jdl/index.js [hunk 3]
```diff
@@ -154,93 +129,85 @@ module.exports = class extends BaseGenerator {
...
-    });
-    return jdlApplications;
+function shouldGenerateApplications(generator) {
+    return !generator.options['ignore-application'] && generator.importState.exportedApplications.length !== 0;
 }
 
 function generateApplicationFiles(generator, application) {
-    const initialWorkingDir = shelljs.pwd();
-    shelljs.cd(path.join(shelljs.pwd().stdout, application.path || ''));
     callSubGenerator(generator, '..', 'app', {
```
Comment (MathieuAA) [inline generators/import-jdl/index.js]: Logically, that's how a subgen should be called (if I'm not wrong). And after looking at the code for the app subgen, I can't understand why it doesn't pick up the .yo-rc.json file. The creation of the files takes place before the subgens are called (if I understand Yeoman precedence rules).

**Comment by jdubois** [comment]
I restarted the build and it passes - should I merge this, so we got it in our upcoming beta release? @MathieuAA are you confident about it?

**Comment by MathieuAA** [comment]
The entity generation works, so there's no regression. The app generation may fail, and I don't want that, even for a Beta. Do you want to release today?

**Comment by jdubois** [comment]
I want to release "soon", but I'm not going to merge the PR if app generation "may fail" - we can still merge later, there is no hurry

**Comment by MathieuAA** [comment]
If soon means before tomorrow, I may be able to make it.

**Comment by jdubois** [comment]
OK, keep me updated, I might have time on Sunday evening for the release

**Comment by MathieuAA** [comment]
Roger that!

**Comment by MathieuAA** [comment]
I've finally found out why it didn't work! Turns out I have to reload the config, which isn't possible for some reasons with `this.config.getAll` (it may not do any I/O, which may be an issue, gotta check). I think I get it working before your beta release Julien.

**Comment by jdubois** [comment]
OK this all look correct - @MathieuAA is it ready to merge? I'll have time today & tomorrow

**Comment by MathieuAA** [comment]
@jdubois except the tests that fail. I'll be honest, I'm not happy with my solution. I'd be ready to merge, as it's a beta --improvable later on, but I'd rather clean up the tests if we keep this _ugly_ solution.

cc @deepu105 knowing you have a lot more experience with Yeoman that I do, do you have an input on this?

File: generators/import-jdl/index.js [hunk 4]
```diff
@@ -154,42 +143,57 @@ module.exports = class extends BaseGenerator {
                     }
                     this.error(`Error while parsing applications and entities from the JDL ${error}`);
                 }
+            }
+        };
+    }
+
+    get install() {
```
Comment (deepu105) [inline generators/import-jdl/index.js]: @MathieuAA not sure about doing the application generation in the install phase. I think it might be better to do the generation in the default phase. Any specific reason for this?
The install phase will queue up along with the install phase of all the sub gen that are composed from this as well and might cause race conditions

File: generators/import-jdl/index.js [hunk 5]
```diff
@@ -201,63 +205,69 @@ module.exports = class extends BaseGenerator {
...
+    }
+    if (application['generator-jhipster'].skipUserManagement) {
+        args.push('--skip-user-management');
+    }
+    if (application['generator-jhipster'].jhiPrefix) {
+        args.push('--jhi-prefix');
+        args.push(application['generator-jhipster'].jhiPrefix);
+    }
+    const done = generator.async();
+    generator.spawnCommand('yo', args).on('close', () => {
```
Comment (deepu105) [inline generators/import-jdl/index.js]: I dont like this part TBH, but its fine for now, i'll rework this when I find time, as its quite tricky to get composing working correctly here

File: generators/import-jdl/index.js (hunk 4 above)
Comment (MathieuAA) [reply generators/import-jdl/index.js]: According to the doc, being in the install scope is recommended when using the spawnCommand

File: generators/import-jdl/index.js (hunk 5 above)
Comment (MathieuAA) [reply generators/import-jdl/index.js]: I'm still trying to get it working when I have the time, but the main issue is passing the context from one generator to the other when composing. I thought it was automatic when composing generators, but I can't get it working.

**Comment by MathieuAA** [comment]
@deepu105 I've managed to make it work without the hack.
I'll update the PR after I make the tests pass.

**Comment by MathieuAA** [comment]
@jdubois Now you can merge it :)

**Comment by jdubois** [comment]
Sorry @MathieuAA but the build badly fails :-(

**Comment by MathieuAA** [comment]
And I can't reproduce it right now

File: generators/app/index.js [hunk 6]
```diff
@@ -130,6 +130,7 @@ module.exports = class extends BaseGenerator {
             defaults: false
         });
 
+        this.availableGeneratorConfig = this.options.availableGeneratorConfig;
```
Comment (deepu105) [inline generators/app/index.js]: @MathieuAA I think this might be duplication of the `configOptions` param that is passed around, can you check?
The purpose of `configOptions` was to pass context around when composing so please do see if you are able to use it without introducing a new param

File: generators/entity/index.js [hunk 7]
```diff
@@ -128,6 +130,12 @@ module.exports = class extends BaseGenerator {
         if (useBlueprint) return;
         return {
             getConfig() {
+                if (this.availableGeneratorConfig) {
+                    this.config = this.availableGeneratorConfig;
+                    this.config.getAll = () => this.config;
```
Comment (deepu105) [inline generators/entity/index.js]: why are these needed?

File: generators/import-jdl/index.js [hunk 8]
```diff
@@ -71,80 +77,69 @@ module.exports = class extends BaseGenerator {
...
-                this.skipClient = this.config.get('skipClient');
-                this.clientFramework = this.config.get('clientFramework');
+                if (!jhiCore.FileUtils.doesFileExist('.yo-rc.json') && !this.generated) {
+                    // we may have to parse the JDL first
+                    return;
+                }
+                let configuration = this.config.getAll();
+                if (!configuration.baseName && jhiCore.FileUtils.doesFileExist('.yo-rc.json')) {
+                    configuration = JSON.parse(fs.readFileSync('.yo-rc.json', { encoding: 'utf-8' }))['generator-jhipster'];
+                    configuration.getAll = () => configuration;
```
Comment (deepu105) [inline generators/import-jdl/index.js]: if such an update is needed, can we do it in a util method in generator-base and reuse it where needed instead of repeating this?

File: generators/app/index.js (hunk 6 above)
Comment (MathieuAA) [reply generators/app/index.js]: roger that, thanks for the tip

File: generators/entity/index.js (hunk 7 above)
Comment (MathieuAA) [reply generators/entity/index.js]: to replace the seemingly absent config passing from one generator to another, but with your previous tip, I might get it to work without it

File: generators/import-jdl/index.js (hunk 8 above)
Comment (MathieuAA) [reply generators/import-jdl/index.js]: yes

**Comment by jdubois** [comment]
@MathieuAA if you are near Ippon we can have lunch together today and go through all this

**Comment by MathieuAA** [comment]
I'm 15mins from Ippon, I believe we can

**Comment by pascalgrimaud** [comment]
@MathieuAA : just tested on my machine, I could reproduce the same issue on Travis.
I'm trying to fix it.

**Comment by MathieuAA** [comment]
@pascalgrimaud You rock

File: generators/entity/index.js (hunk 7 above)
Comment (pascalgrimaud) [reply generators/entity/index.js]: @MathieuAA : these 5 lines made `this.config` wrong. If I deleted these 5 lines, it seems to work

**Comment by pascalgrimaud** [comment]
@MathieuAA : my use case to reproduce the issue:
- copy an existing `.yo-rc.json` into a new folder
- copy some entities inside `.jhipster/` folder, for example Operation, BankAccount and Label
- launch `jhipster --with-entities`

I wonder if the new variable you introduced `this.availableGeneratorConfig` is really usefull ?

**Comment by MathieuAA** [comment]
@pascalgrimaud Not if we replace it by `this.configOptions`, which seems to be the convention for this use case.

**Comment by pascalgrimaud** [comment]
Here the commit: https://github.com/pascalgrimaud/generator-jhipster/commit/89df4207e437b0ec754fd00eb21ca30388b42e05 -> can you check if it's ok for you ? If yes, you can take the commit, or I can do a PR against your branch

Here the current Travis builds: https://travis-ci.org/pascalgrimaud/generator-jhipster

But to be honest, I don't really understand what I'm doing :-D

**Comment by MathieuAA** [comment]
If there's a `--with-entities`, that's one more use case, as I won't have to call the entity subgen (the app subgen will)

**Comment by MathieuAA** [comment]
And the app generation + entity generation work?

**Comment by pascalgrimaud** [comment]
I only tested the default use case in Travis.
I didn't test the new feature app+entity in same file.

**Comment by MathieuAA** [comment]
I'm going to merge this and test later. Thanks a lot Pascal :)

**Comment by MathieuAA** [comment]
I'm going to merge this and test later. Thanks a lot Pascal :)

**Comment by jdubois** [comment]
Merged! Let's do this :-)
